lib/Validator/Constraints/FieldDefinitionDefaultValueValidator.php

```diff
@@ -51,4 +52,56 @@ protected function generatePropertyPath($errorIndex, $errorTarget)
     {
         return 'defaultValue';
     }
+
+    /**
+     * Returns the field definition $value refers to.
+     * FieldDefinition object is needed to validate field value against field settings.
+     *
+     * @param FieldData|ValueObject $value ValueObject holding the field value to validate, e.g. FieldDefinitionData.
+     *
+     * @throws \InvalidArgumentException If field definition cannot be retrieved.
+     *
+     * @return \eZ\Publish\API\Repository\Values\ContentType\FieldDefinition
+     */
+    protected function getFieldDefinition(ValueObject $value)
+    {
+        return $this->getUpdatedFieldDefinition($value);
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\ValueObject $value
+     *
+     * @return \eZ\Publish\Core\Repository\Values\ContentType\FieldDefinition
+     */
+    private function getUpdatedFieldDefinition(ValueObject $value): FieldDefinition
+    {
+        $fieldDefinition = $value->fieldDefinition;
+        $properties = $this->objectToArray($fieldDefinition);
+        $properties['fieldSettings'] = $value->fieldSettings;
+        $updatedFieldDefinition = new FieldDefinition($properties);
+
+        return $updatedFieldDefinition;
+    }
+
+    /**
+     * @param $object
+     *
+     * @return array
+     */
+    private function objectToArray($object): array
+    {
+        $clone = (array)$object;
+        $rtn = array();
```

Comment:
short array syntax, 
still, don't know how this is nicer than reflection.

Reply:
Also, not fun of $rtn variable name and I think getUpdatedFieldDefinition() deserves proper doc with explanation.

Reply:
> still, don't know how this is nicer than reflection.

- Performance. Reflections are slow. They can be used in tests (if there's no other way around), never should be used in production code. 
- Reflection is needed when you need to either access private/protected properties or get knowledge on how the object is build. The need for that contradicts DDD and indicates some architecture issues (which indeed we do have)

Reply:
I know all of that, but tbh I believe that performance won't be an issue here.
And the second point is exactly what this whole thing is about - accessing private property :D